Adds summary of potentials (Fixes issue #246)

hey-mathis opened the pull request:
Fixes # 246.

Adds Snakemake rule that generates .nc and .csv files (longform data) that summarize the potentials (= per-tech constriants) for each technology and each location.

## Checklist

Any checks which are not relevant to the PR can be pre-checked by the PR creator. All others should be checked by the reviewer. You can add extra checklist items here if required by the PR.

- [x] CHANGELOG updated
- [x] Minimal workflow tests pass
- [x] Tests added to cover contribution
- [x] Documentation updated
- [x] Configuration schema updated

timtroendle commented on scripts/summarize_potentials.py:
```diff
@@ -0,0 +1,56 @@
+    for potential in list_of_potentials:
+        aux = model.get_formatted_array(potential)
+        for tech in list_of_techs:
+            for loc in list_of_locs: 
```
This implementation is a little risky as "for" loops are notoroiously slow in interpreted languages like Python. It should be fine here because there are not that many iterations (max ~500 locs * ~20 techs = ~10000 iterations).

timtroendle commented on scripts/summarize_potentials.py:
```diff
@@ -0,0 +1,56 @@
+
+
+def summarize_potentials(
+    path_to_model, path_to_output_csv, path_to_output_netcdf, sf
```
Please add [type hints](https://docs.python.org/3/library/typing.html) to the function arguments. This allows people and machines to better understand what's going on.

Frankly, we haven't done that much in the past but we are intending to always do that moving forward. For an example, have a look at this script:
https://github.com/calliope-project/euro-calliope/blob/92a93dfc2bd7f29f6745372bc45c5013234753af/scripts/nuclear/regional_capacity.py#L8

hey-mathis commented:
Fixed the suggested changes:
- Changed file names from _ to - (Now: "summary-of-potentials.nc")
- Improved code style using Flake8 and EditorConfig
- Merged the comment
- Kept the for loops
- Added type hints
- Reexecuted the minimal workflow tests


Attached the resulting files:

National:
[summary-of-potentials.csv](https://github.com/calliope-project/euro-calliope/files/10269406/summary-of-potentials.csv)

Regional:
[summary-of-potentials.csv](https://github.com/calliope-project/euro-calliope/files/10269415/summary-of-potentials.csv)

Continental:
[summary-of-potentials.csv](https://github.com/calliope-project/euro-calliope/files/10269417/summary-of-potentials.csv)

timtroendle commented on scripts/summarize_potentials.py:
```diff
@@ -0,0 +1,77 @@
+
+    summary = np.empty((len(list_of_techs), len(list_of_potentials), len(list_of_locs), 2), dtype='<U32')
+    summary[:] = np.nan
+    summary = xr.DataArray(summary, dims=("techs", "potentials", "locs", "value_unit"),
```
Modeling value and unit as a single dimension is rather unusual. In fact, neither of the two is a dimension. The data is three-dimensional: techs, locs, potentials. This also leads to an usual structure of the csv.

You _can_ model units as a ["non-dimensional coordinate"](https://docs.xarray.dev/en/stable/user-guide/data-structures.html#coordinates) if you want. I think this would be the cleanest way of modeling it. But you don't have to.

timtroendle reviewed (changes requested):
Both the csv and nc have an usual structure with the way they handle value and units. The longform of the csv is unusal in that usually you wouldn't stack unit and value. This is difficult to handle, most importantly because the final column (which does not have a name) as different types: floats and strings.

Rather unstack the data so that the csv has a value column and a unit column.

Also, I'd prefer to drop nans. There will be too many of them and in most cases they do not have a meaning.

hey-mathis commented:
Fixed the suggested changes:

- Modeled the units as non-dimensional coordinate. Therefore we are back to three dimensions: techs, locs, and potentials. This leads to an 'unstacked' and more natural .csv file.
- Dropped the NaNs: Therefore, a potential/tech/location combination does not appear in the csv file if (a) the technology is not installed at the given location, or (b) the constraint variable is not defined for the given technology, or (c) the constraint variable is not set for the given technology at the given location. However, when the constraint variables are set to "0.0" or "inf" the rows are retained.
- Adapted comments
- Reexecuted all minimal workflow tests

Attached, the resulting csv files (.nc files are not supported in this chat, let me know, if you need them):

National: [summary-of-potentials.csv](https://github.com/calliope-project/euro-calliope/files/10332350/summary-of-potentials.csv)

Regional: [summary-of-potentials.csv](https://github.com/calliope-project/euro-calliope/files/10332351/summary-of-potentials.csv)

Continental: [summary-of-potentials.csv](https://github.com/calliope-project/euro-calliope/files/10332353/summary-of-potentials.csv)

timtroendle commented on scripts/summarize_potentials.py:
```diff
@@ -0,0 +1,84 @@
+
+    summary = np.empty((len(list_of_techs), len(list_of_potentials), len(list_of_locs)))
+    summary[:] = np.nan
+    units = np.empty((len(list_of_techs), len(list_of_potentials), len(list_of_locs)), dtype='<U32')
```
Shouldn't `units` be of "dimension" `list_of_potentials` only? Unit is defined per potential.

timtroendle commented on scripts/summarize_potentials.py:
```diff
@@ -0,0 +1,84 @@
+                           coords={"techs": list_of_techs,
+                                   "potentials": list_of_potentials,
+                                   "locs": list_of_locs,
+                                   "unit": (["techs", "potentials", "locs"], units)}
```
As above, shouldn't this be potentials only?

timtroendle commented on scripts/summarize_potentials.py:
```diff
@@ -0,0 +1,84 @@
+                    summary.loc[dict(techs=tech, locs=loc, potentials=potential)] = aux.loc[
+                        dict(techs=tech, locs=loc)] / considered_potentials[potential]["sf"]
+                    summary.coords["unit"].loc[
+                        dict(techs=tech, locs=loc, potentials=potential)] = considered_potentials[potential]["unit"]
```
As above. This could then move two `for` loops up. It's simply `dict(potentials=potential)`

timtroendle commented on scripts/summarize_potentials.py:
```diff
@@ -0,0 +1,84 @@
+                        dict(techs=tech, locs=loc, potentials=potential)] = considered_potentials[potential]["unit"]
+                except KeyError:
+                    # If a technology "tech" is not installed in location "loc", or the potential "potential"
+                    # is not defined for "tech" at "loc", value and unit remain "NaN".
```
As above. Unit would then never be NaN.

timtroendle commented on scripts/summarize_potentials.py:
```diff
@@ -0,0 +1,84 @@
+import xarray as xr
+
+
+def summarize_potentials(
```
In the file-, function-, and rule-name: can you use British English `summarise`?

timtroendle reviewed (changes requested):
The file looks great now. Very helpful and it may already have identified a bug (#252).

Two more small changes for the code. The first (language) is necessary, the second (dimensions of unit) not. Give the second a try and if you cannot fix it quickly, don't bother.

No need to rerun the minimal workflow.

timtroendle reviewed (changes requested):
Thanks for the latest changes. I was about to accept this PR when I figured there is still something missing while ticking the checkboxes: documentation! 🤦‍♂️ Sorry, my bad!

We need to document these changes in the following file of the documentation:
https://github.com/calliope-project/euro-calliope/blob/develop/docs/model/overview.md

https://euro-calliope.readthedocs.io/en/latest/model/overview/

hey-mathis commented:
Alright, I fixed the following issues: 

1) Changed "summarize" to "summarise" in file-, function-, and rule name 
2) Changed dimension of "unit" to len(list_of_potentials) -- great idea and straight-forward to implement, reduces number of loops ...
3) Updated folder structure in the documentation file:
- Added "summary-of-potentials.csv" and an explanation
- Added "summary-of-potentials.nc" and an explanation
- Added "locations.csv" and an explanation
- Changed: {technology-group}.csv -> {technology-group}.yaml (These are .yaml files, not .csv files, right?)

Don't worry, let me know, if you spot anything else :)

timtroendle reviewed (approved):
Awesome, thanks!
